pantheon-bundle/src/main/java/com/redhat/pantheon/servlet/assets/ImageServletFilter.java

```diff
@@ -0,0 +1,57 @@
+package com.redhat.pantheon.servlet.assets;
+
+import org.apache.sling.servlets.annotations.SlingServletFilter;
+import org.osgi.service.component.annotations.Component;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.util.Base64;
+
+import static com.redhat.pantheon.conf.GlobalConfig.IMAGE_PATH_PREFIX;
+
+/**
+ * A servlet filter which serves image assets only. Images must be referened using an absolute path comprised
+ * of a prefix (see @{link IMAGE_PATH_PREFIX} for the exact value) plus an identifier. In this servlet the identifier is
+ * a Base64 encoding of the actual path stored in the JCR repository for the image.
+ * </br>
+ * </br>
+ * So for example, an image path might be:
+ * http://my.server/imageassets/L2NvbnRlbnQvcmVwb3NpdG9yaWVzL3BhbnRoZW9uU2FtcGxlUmVwby9pbWFnZXMvdGlnZXIucG5n
+ * with /imageassets being the prefix and everything after (exluding the separating slash) being a Base64 encoding of the
+ * path: /content/repositories/pantheonSampleRepo/images/tiger.png which resolves absolutely and unambigously to an image
+ * in the system.
+ *
+ * @author Carlos Munoz
+ */
+@Component(
+        service = Filter.class
+)
+@SlingServletFilter(
+        methods = "GET",
+        pattern = "/imageassets/.*")
```

Comment:
I'm still not wild about using a Filter *as* a servlet, but this does appear to be the only way to do it. I tested a regular servlet with `@SlingServletPaths(value = "/pantheon/ben/.*")` and it did not behave the way I hoped. I had to literally make a request to `/pantheon/ben/.*` to execute the servlet - requesting `/pantheon/ben/anything` produced a 404.

One other possibility - we could write our own `DefaultErrorHandlerServlet` that is able to check for cases where the request starts with `/imageassets/` and do something clever there. We could also extend the servlet to check for things like `/modules/by-uuid/` and do something different. Not sure that that's any better really... but might be worth debating the pros and cons.

Reply:
This is a restriction in Sling's implementation of the servlet spec. In other implementations like Tomcat or Jetty you are allowed to map servlets to wildcard patterns.

Your suggested approach seems similar to the Filter approach, just using a different mechanism (an error handler). In which case I'd argue a filter is still a better approach. But I admit I still don't quite understand your concern about using a filter. I actually see this as a perfect example of how it should be used. The server is capturing a request, modifying it and then sending it the right way.

Reply:
The "sending it the right way" part is what I've got a problem with. In my mind, a request ought to be handled successfully (albeit perhaps not perfectly) even if you remove all filters from the system. So filters are "nice to have" not necessarily "critical to have." In this case, removing the filter would cause the request to fail outright.

I agree that this is the best we can do in Sling, and that the Error Handler approach isn't "better," just "different." But I do think that logically it makes more sense. Sling is throwing up its hands and saying "I don't know what to do with this, deal with it somehow" and then we deal with it however we like.

I don't feel strongly enough about it to ask for a change, but it's something we can keep in mind if circumstances change in the future, such as if we need to implement another filter for some other feature and they wind up conflicting with each other.